toolbox/src/base/ConvertStdVector.cpp

```diff
@@ -0,0 +1,123 @@
+                                                 std::vector<double>& output);
+// Bool to other numeric types
+template void wbt::convertStdVector<bool, bool>(const std::vector<bool>& input,
+                                                std::vector<bool>& output);
+template void wbt::convertStdVector<bool, int>(const std::vector<bool>& input,
+                                               std::vector<int>& output);
+template void wbt::convertStdVector<bool, double>(const std::vector<bool>& input,
+                                                  std::vector<double>& output);
+// Double to other numeric types
+template void wbt::convertStdVector<double, double>(const std::vector<double>& input,
+                                                    std::vector<double>& output);
+template void wbt::convertStdVector<double, int>(const std::vector<double>& input,
+                                                 std::vector<int>& output);
+template void wbt::convertStdVector<double, bool>(const std::vector<double>& input,
+                                                  std::vector<bool>& output);
+// String to string
+template void wbt::convertStdVector<std::string, std::string>(const std::vector<std::string>& input,
+                                                              std::vector<std::string>& output);
+
+// Explicit specialization for std::string type
+// ============================================
+
+template <>
+void wbt::convertStdVector<std::string, int>(const std::vector<std::string>& input,
+                                             std::vector<int>& output)
+{
+    output.clear();
+    output.resize(input.size());
+    std::transform(input.begin(), input.end(), output.begin(), [](const std::string& str) {
+        return std::stoi(str);
+    });
+}
+
+template <>
+void wbt::convertStdVector<std::string, bool>(const std::vector<std::string>& input,
+                                              std::vector<bool>& output)
+{
+    output.clear();
+    output.resize(input.size());
+    std::transform(input.begin(), input.end(), output.begin(), [](const std::string& str) {
+        return static_cast<bool>(std::stoi(str));
+    });
+}
+
+template <>
+void wbt::convertStdVector<std::string, double>(const std::vector<std::string>& input,
+                                                std::vector<double>& output)
+{
+    output.clear();
+    output.resize(input.size());
+    std::transform(input.begin(), input.end(), output.begin(), [](const std::string& str) {
+        return std::stod(str);
+    });
+}
+
+template <>
+void wbt::convertStdVector<int, std::string>(const std::vector<int>& input,
+                                             std::vector<std::string>& output)
+{
+    output.clear();
+    output.resize(input.size());
+    std::transform(input.begin(), input.end(), output.begin(), [](const int& num) {
+        return std::to_string(num);
```

Comment:
How are you using this methods? If you are using this methods to convert parameters in configuration files, typically you do not want this conversion to be influenced by the locale setting, and so `to_string` is not the appropriate functions. I suspect also `std::stod` is affected. 

I originally became aware of this edge case in the context of URDF parsers, see https://github.com/robotology/idyntree/issues/288 . Most of the time this kind of problems it not recognized because:
* The default locale setting for C/C++ program is `"C"`. 
* The user-preferred language for most developers is `en-US`, or in general `en-SOMETHING`. 

However, if the libraries is used in a process that uses the Qt libraries on a computer in which the user-preferred locale is something like `it-IT` or `es-ES`, the Qt libraries set the C/C++ locale to be the user-preferred locale, and the libraries stop working properly. As you may imagine, this is not a behavior that is easy to debug.

Reply:
> As you may imagine, this is not a behavior that is easy to debug.

I remember "piovere madonne" from that issue :sweat_smile: In this case no file is involved, at least, not at the current stage.

This template is used inside `wbt::Parameters` for converting vector parameters between different data types. Values are usually read from the block's mask in Simulink.

I don't see how this can affect `std::to_string`, but I see the point of the other way around, and I mean `std::stod` and `std::stoi` which I use in many points. The origin of this feature is to give the possibility of calling:

```cpp
// Store a parameter as a double
Parameter p(42, {ParameterType::DOUBLE, 0, 1, 1, "ParamName"});

// Insert it in a container with other parameters
Parameters params;
params.storeParameter(p);
...

// Get the parameter that is cast internally
int value;
getParameter("ParamName", value);
```

This should hold also for `std::string`.

In future versions I plan to give the possibility of override those variables for give flexibility to the automatic code generation. In this case, probably the parameters to override will be stored in a configuration file.

@traversaro Do you think this problem should be addressed by whom will parse the configuration? It will not be responsibility of the toolbox. And, if this is the case, all `RFModule` should be careful of this behavior (how does yarp handle it? I guess it does not).

Reply:
> And, if this is the case, all RFModule should be careful of this behavior (how does yarp handle it? I guess it does not).

I do not think that `yarp::os::Property` use any kind of double to string facilities provided by C++. 

> I don't see how this can affect std::to_string, but I see the point of the other way around

Basically, you risk to store doubles as `10,000`. Depending on the usage, it may or not may be a problem. 

> Values are usually read from the block's mask in Simulink.

They are read as strings or directly as doubles? In the generated code, how they are stored? If they are read or stored as strings, I think we already have the problem, as a user would but `0.10` in its Simulink mask, and in could be read if the wrong way depending on the locale.

Reply:
> They are read as strings or directly as doubles?

Numeric types are read as `double` and strings as `string`s. Simulink functions are used, and we do not have control on them. Hopefully they are robust to locale.

> In the generated code, how they are stored? 

About serialization / deserialization to the RTW file (which is kind of a database read by the automatic generation pipeline) the following logic is used:

- Strings are serialized as `"stringContent"`
- Numeric types are all converted to `double` and serialized as `42.0`

Again, I use Simulink functions for doing this.

If what I wrote so far is right, the parameters are correctly read by Simulink, correctly serialized to the RTW file, and correctly read from the RTW file from the autogenerated code pipeline.

### Possible problems I see

These helpers are used only when the block developer wants for some reason to get a parameter with a type different than the original type (specified in the parameter metadata). This step might be locale-dependent.

#### Numeric -> string

If you want to convert a double to a string you will get `"10,000"` as you mentioned. But, since this is already kind of downstream, after the cast unlikely you will compare numeric types as strings. Rather than converting the numeric parameter to string you would rather convert the other string to numeric and then compare the numeric values.

#### String -> numeric

Problems would arise if the block developer chooses to store a numeric type as a string (non-default behavior). In this case I have no idea what would be the result of calling `std::stod("42.0")` on locales that assume the `"42,0"` as a separator. Anyway, this is a very edge case.

### Final thoughts

In the light of these considerations, the current situation doesn't look problematic. Does it sound good even for you @traversaro?

Reply:
Sorry, I think I finally got it. 

The `Parameters` class is a generic key-value storage in which values are typed (either `std::string`, `bool`, `int` or `double`) used to set parameters for code-generated blocks. 
However, even if the parameters are typed, the class offers a way of converting internally parameters from one type to another. However, this feature is not used currently in any part of the code,  and at the moment we are not currently plan to use it, right? 

At this point, the only thing I am afraid is the current situation:
* a given block takes in input a `double` parameter, but for some reason the external C++ code save it as a string parameter, with some strange locale setting
* if I understand correctly the code in https://github.com/robotology/wb-toolbox/pull/77/files/05bd751d6d557cb110d1437bb845ba4a7812fcd0#diff-169c37a45eec0ad411c3e3a6f9052cf3R208, the parameter parsing is successful, but the parameter is actually converted implicitly from string to double. 

I must say that if I got the big picture, I do not understand the need for the implicit conversion functionality in `Parameters`, but now that it is there I do not think it make sense to try to remove it. 

I have nevertheless two suggestions, but I am not sure if they make sense:
* Add a line in `Parameters`'s documentation to say that for implicit conversions between std::string and the other types the standard functions `std::to_string`, `std::stoi`, etc etc are used, and the behavior of this functions are influenced by the process-global behavior.
* If it is not too complicated, consider also the implicit type when performing the check on the parameters.

Reply:
> However, this feature is not used currently in any part of the code, and at the moment we are not currently plan to use it, right?

We do not currently use it, but on the long run I thought it was a nice addition especially if we split the core of the toolbox. Consider the following snippet:

```cpp
Parameters parameters;
parameters.storeParameter(42, {ParameterType::DOUBLE,
                               /*index=*/0,
                               /*rows=*/1,
                               /*cols=*/1,
                               "paramName"});
```

The type specified in the metadata **always** wins. The usage without conversion might have led to a template instantiation difficult to figure out. In this way we don't mind.